src/map/handler/Map.TapHold.js

```diff
@@ -0,0 +1,108 @@
+import {Map} from '../Map';
+import {Handler} from '../../core/Handler';
+import * as DomEvent from '../../dom/DomEvent';
+import {Point} from '../../geometry/Point';
+import * as Util from '../../core/Util';
+import * as Browser from '../../core/Browser';
+
+/*
+ * L.Map.TapHold is used to simulate `contextmenu` event on long hold,
+ * which otherwise is not fired by mobile Safari.
+ */
+
+var tapHoldDelay = 600;
+
+// @namespace Map
+// @section Interaction Options
+Map.mergeOptions({
+	// @section Touch interaction options
+	// @option tapHold: Boolean
+	// Enables simulation of `contextmenu` event, default is `true` for mobile Safari.
+	tapHold: Browser.touch && Browser.safari && Browser.mobile,
+
+	// @option tapTolerance: Number = 15
+	// The max number of pixels a user can shift his finger during touch
+	// for it to be considered a valid tap.
+	tapTolerance: 15
+});
+
+export var TapHold = Handler.extend({
+	addHooks: function () {
+		DomEvent.on(this._map._container, 'touchstart', this._onDown, this);
+	},
+
+	removeHooks: function () {
+		DomEvent.off(this._map._container, 'touchstart', this._onDown, this);
+	},
+
+	_onDown: function (e) {
+		if ('pointerType' in e && e.pointerType === (e.MSPOINTER_TYPE_MOUSE || 'mouse')) {
+			return; // workaround #7058
+		}
+		clearTimeout(this._holdTimeout);
+		if (e.touches.length !== 1) { return; }
+
+		var first = e.touches[0];
+		this._startPos = this._newPos = new Point(first.clientX, first.clientY);
+
+		this._holdTimeout = setTimeout(Util.bind(function () {
+			this._cancel();
+			if (!this._isTapValid()) { return; }
+
+			// prevent simulated mouse events https://w3c.github.io/touch-events/#mouse-events
+			// Note: direct use of addEventListener is temporary workaround for #7029/3
+			document.addEventListener('touchend', DomEvent.preventDefault, false);
+			document.addEventListener('touchend', this._cancelClickPrevent, false);
+			document.addEventListener('touchcancel', this._cancelClickPrevent, false);
+			this._simulateEvent('contextmenu', first);
+		}, this), tapHoldDelay);
+
+		DomEvent.on(document, 'touchend touchcancel contextmenu', this._cancel, this);
+		DomEvent.on(document, 'touchmove', this._onMove, this);
+	},
+
+	_cancelClickPrevent: function cancelClickPrevent() {
+		document.removeEventListener('touchend', DomEvent.preventDefault, false);
+		document.removeEventListener('touchend', cancelClickPrevent, false);
+		document.removeEventListener('touchcancel', cancelClickPrevent, false);
+	},
```

Comment:
No, with your change we loose context.
So we had 2 options: either bind context, or use named function.

Reply:
Hmm ok. Why are you not using `DomEvent` here? Then it would be possible to use the same context right?

Reply:
Because of #7077. See 3a90acf1536176f72e2d738940284e4668718f47

Reply:
Well, #7077 is not in case anymore, but I still prefer to keep native add/removeEventListener here to emphasize that only native touch events are relevant here.
(Will update comments)